### ISSUE TITLE
GetOpenFileName

### ISSUE THREAD

**Issue by MikeYeager** [microsoft/CsWin32#191]
I'm pretty new to this, but I have some code using DllImport to call GetOpenFileName that works and I wanted to see if I could do it with CsWin32. The OPENFILENAMEW struct isn't very C# friendly, but shouldn't something like this work? The dialog doesn't even open.

```
        public static string MyGetOpenFileName()
        {
            var fileName = string.Empty;

            unsafe
            {
                fixed (char* file = new char[100])
                {
                    var request = new OPENFILENAMEW
                    {
                        lStructSize = (uint)sizeof(OPENFILENAMEW),

                        //lpstrInitialDir = null, 
                        //lpstrFilter = null,
                        //lpstrDefExt = null,

                        lpstrFile = new PWSTR(file),
                        nMaxFile = 100,

                        //hwndOwner = new HWND(),
                        //hInstance = new HINSTANCE(),
                        //Flags = 0,
                        //FlagsEx = 0,
                        //lpstrCustomFilter = null,
                        //lpstrTitle = null,
                        //lCustData = new LPARAM(),
                        //lpfnHook = null,
                        //lpstrFileTitle = null,
                        //lpTemplateName = null,
                        //nFileExtension = 0,
                        //nFileOffset = 0,
                        //nFilterIndex = 0,
                        //nMaxCustFilter = 0,
                        //nMaxFileTitle = 0,
                        //dwReserved = 0,
                        //pvReserved = null,
                    };

                    var s = request.lStructSize;

                    if (PInvoke.GetOpenFileName(ref request))
                        fileName = new string(request.lpstrFile);
                }
            }

            return fileName;
        }
```

**Comment by AArnott** [comment]
Yes, something like that should work. I don't know why it doesn't. Perhaps you're not initializing enough of the struct.

**Comment by MikeYeager** [comment]
I've uncommented the initialization, tried initializing with null, tried initializing with a default [new PWSTR(), etc.]. I've tried not initializing dwReserved and pvReserved (my ImportDll code doesn't initialize them). I guess I'll review the Win32 docs again keep trying. Thanks.

**Comment by jnm2** [comment]
@MikeYeager My approach would be to copy all the source generated code to a file you control, then slowly replace parts of it with your original handwritten declarations bit by bit until you find the piece that causes it to start working. (Or in reverse, till you find the piece that causes it to start failing.)

**Comment by AArnott** [comment]
If you do find something broken with the codegen, please let us know.

**Comment by MikeYeager** [comment]
@jnm2 Thanks for the tip, but I have already tried to go down that road. It seems the problem (from what I can tell) is with the OPENFILENAMEW struct that is generated, not the call itself. The types differ from those the GetOpenFileName dialog uses. I tried creating my own copy of OPENFILENAMEW, however types in that class like HWND, HINSTANCE, PWSTR, etc. are not accessible.

@AArnott So far, I've been unable to the generated code to work. Below is a screen shot of the OPENFILENAMEW struct created by VsWin32 and the code for the struct I've been successfully using to PInvoke the dialog manually. The .NET types differ quite a bit, but I can't figure out how to investigate the types listed above to find the issue.
![image](https://user-images.githubusercontent.com/7063813/111343040-a63d5b80-8640-11eb-820c-096c6f8af65d.png)

**Comment by jnm2** [comment]
> however types in that class like HWND, HINSTANCE, PWSTR, etc. are not accessible.

They should be accessible; just copy them over from the source generator into your own code as part of the process.

**Comment by AArnott** [comment]
I don't see anything incompatibly different between the two structs.

**Comment by MikeYeager** [comment]
@jnm2 When I view the generated structs those types I cannot navigate to their definitions. Notice in the screen shot above that those types are not syntax colored. Perhaps there is another way to find that generated code? Any help would be appreciated. Thank you.

**Comment by AArnott** [comment]
@mikeyeager you can use Ctrl-T on those type refs to jump to the files that define them.
Also if you're using VS 16.9, you can expand the References->Analyzers node to find _all_ generated code.

**Comment by MikeYeager** [comment]
@AArnott I didn't see anything incompatible at first glance, but I don't yet completely understand how the generated HWND HINSTANCE, PWSTR, etc. types work. These seem to be custom .NET types created by the generator since I can 'new' them up. When I navigate to their definitions that way, I see for example, a partial struct that uses some new C#9 language features like the nint type. I can't tell if I need to worry about what else might be present due to the partial keyword. What I know about nint for example is that it *should* be compatible with an IntPtr. I'm still looking through it all, but no luck getting a working example with generated code. I'm now trying to dump an instance of each struct to binary files so I can compare them that way.

**Comment by AArnott** [comment]
`nint` is literally the same thing at runtime as `IntPtr`. It's just C# syntax sugar.
I'll try to find something as well.

**Comment by MikeYeager** [comment]
@AArnott References -> Analyzers. Brilliant! Thank you.

**Comment by AArnott** [comment]
I added this code to your sample:

```cs
            Console.WriteLine(PInvoke.CommDlgExtendedError());
```

This printed `1`, which is documented [here](https://docs.microsoft.com/en-us/windows/win32/api/commdlg/nf-commdlg-commdlgextendederror) as `CDERR_STRUCTSIZE`. So apparently the `sizeof` that you're using is wrong.
It's set to 140. What is it for your working version?

**Comment by AArnott** [comment]
152 is the right size. The `Pack=1` value is where the bug is. It shouldn't be there. This is a bug in the metadata.
